test/Base_Tests/src/Semantic/Multi_Value_As_Type_Refinement_Spec.enso

```diff
@@ -0,0 +1,272 @@
+            ab = make_a_and_b
+            r = case ab of
+                A.A_Ctor x -> "matched: "+x.to_text
+                _ -> "structural matching of A.A_Ctor failed"
+            r.should_equal "matched: 1"
+
+        group_builder.specify "should structural matching match other types?" pending="TODO: decide if we keep this test inhttps://github.com/enso-org/enso/issues/12142" <|
+            ab = make_a_and_b
+            r = case ab of
+                B.B_Ctor x -> "matched: "+x.to_text
+                _ -> "structural matching of B.B_Ctor failed"
+            r.should_equal "matched: (A_Ctor 1)"
+
+        group_builder.specify "calling a method on one of the types should not lose the intersection type" pending="TODO: https://github.com/enso-org/enso/issues/12143" <|
+            ab = make_a_and_b
+
+            # Checked variant can hide the B part
+            ab.a_id . is_a A . should_be_true
+            ab.a_id . is_a B . should_be_false
+            # But it can be uncovered via explicit cast
+            b = (ab.a_id):B
+            b.is_a A . should_be_false
+            b.is_a B . should_be_true
+
+            new_ab = (ab.a_id):A&B
+            new_ab.is_a A . should_be_true
+            new_ab.is_a B . should_be_true
+            new_ab.a_method . should_equal "A method"
+            new_ab.b_method . should_equal "B method"
+
+            # But unchecked variant should keep both types and not hide anything
+            ab.a_id_unchecked . is_a A . should_be_true
+            #ab.a_id_unchecked . is_a B . should_be_true
+            ab.a_id_unchecked.a_method . should_equal "A method"
+            #ab.a_id_unchecked.b_method . should_equal "B method"
+
+            # The same should apply to the B part
+            ab.b_id . is_a A . should_be_false
+            ab.b_id . is_a B . should_be_true
+            new_ab_2 = (ab.b_id):A&B
+            new_ab_2.is_a A . should_be_true
+            new_ab_2.is_a B . should_be_true
+            ab.b_id_unchecked . is_a A . should_be_true
+            ab.b_id_unchecked . is_a B . should_be_true
+            ab.b_id_unchecked.a_method . should_equal "A method"
+            ab.b_id_unchecked.b_method . should_equal "B method"
+
+        group_builder.specify "calling `.catch` on an intersection type should not lose the refinements" <|
+            ab = make_a_and_b
+            r = ab.catch Any _->"catched"
+            r.is_a A . should_be_true
+            r.is_a B . should_be_true
+            r.a_method . should_equal "A method"
+            r.b_method . should_equal "B method"
+
+        group_builder.specify "calling `.throw_on_warning` on an intersection type should not lose the refinements" <|
+            ab = make_a_and_b
+            r = ab.throw_on_warning
+            r.is_a A . should_be_true
+            r.is_a B . should_be_true
+            r.a_method . should_equal "A method"
+            r.b_method . should_equal "B method"
```

Comment:
This was supposed to be an example for the #12143 bug. But apparently this test is already passing?!

I'm really surprised. This was supposed to be a minimal repro of the example that @JaroslavTulach was working-around in commit 9bb73aec68313e83169f29dd785604a5b933d00c.

@JaroslavTulach any idea why you needed a workaround in the Column code but here it seems to work as expected?

I'm a bit anxious that the behaviour seems to be slightly random - in some contexts the intersection is retained and in some it is lost. That is concerning.

I'd be happy if we can fix that, and ideally before - find a repro that actually fails a similar test and add it to the suite.

Reply:
After all I've found a repro: b8407da